---
title: "manuscript"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  word_document:
    reference_docx: "paper-files/style_ref.docx"
  pdf_document:
    keep_tex: yes
    fig_caption: yes
    latex_engine: pdflatex
    fig_width: 7
    fig_height: 6
  html_document:
    df_print: paged
bibliography: "paper-files/citations.bib"
csl: "paper-files/apa.csl"
header-includes:
- \usepackage[left]{lineno}
- \linenumbers
- \usepackage{setspace}\doublespacing
- \usepackage{sectsty}\sectionfont{\fontsize{12}{12}\selectfont}
- \usepackage{sectsty}\subsectionfont{\normalfont\itshape\fontsize{12}{12}\selectfont}
- \usepackage[round]{natbib}
indent: yes
sansfont: Times New Roman
fontsize: 12pt
always_allow_html: yes
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(snpR); library(data.table); library(ggplot2); library(cowplot)
```

# Abstract

Population geneticists use a wide range of approaches to measure population genetic diversity. While measures which incorporate allele frequencies (such as observed and expected heterozygosity, nucleotide diversity, Tajima's D, and so on) are well developed for single nucleotide polymorphism (SNP) data, there are only a few allele frequency independent estimators which are broadly in use. The number of segregating sites (or SNPs) is one of the more prevalent of these; however, unlike allelic richness or private allele frequencies, segregating site counts are rarely adjusted to correct for unequal sample sizes or missing data rates between populations. We here propose a corrected estimator for the number of segregating sites expected in a population following rarefaction, which allows for an unbiased approach for comparing the number of segregating sites between multiple populations.

# Introduction

Estimating the degree of genetic diversity has long been a critically important and widespread practice in conservation genetics [@dewoodyLongstandingSignificanceGenetic2021]. Genetic diversity, or the diversity alleles, genotypes, and segregating loci within a population [@frankhamIntroductionConservationGenetics2002], is the both the basic foundation upon which selection acts and thus in part controls population adaptability [@Kardos2021], and is strongly correlated with fitness both at the individual and population level, even when no obvious functional ties are known [reviewed in @dewoodyLongstandingSignificanceGenetic2021]. 

Estimators of population genetic diversity are themselves extremely diverse, but essentially fall into two categories: those that are depend on allele frequencies and those that do not. Allele frequency dependent statistics, such as expected heterozygosity ($H_{e}$), the F-statistics ($F_{IS}$, $F_{ST}$, and $F_{IT}$)[@Weir1984], nucleotide diversity ($\pi$), and Tajima's $\theta$ [@Tajima1989] all rely principally on the frequency of the alleles across any surveyed loci; whereas allele frequency independent statistics, such as Watterson's $\theta$ [@wattersonNumberSegregatingSites1975], counts of alleles per loci, the number of observed segregating sites, and private allele counts do not. This is not a trivial difference:  evolutionary forces do not act equally on the number of loci/alleles in a population and the frequency of those alleles. Tajima's D, for example, is a powerful and broadly used statistic fundamentally based on the difference between Tajima's $\theta$ and Watterson's $\theta$ that can be used to both detect selection and population demographic changes [@Tajima1989] due to the way that those forces act on the balance of allele frequencies within populations.

The average allele count per locus (which is usually corrected to allelic richness [@kalinowskiCountingAllelesRarefaction2004]) has historically been one of the most prevalent allele frequency independent measures of genetic diversity. While this is a particularly useful measure in microsatellite and other datasets utilizing heavily polyallelic markers, it is less so in single-nucleotide polymorphism (SNP) datasets. SNP datasets are usually biallelic by design, and thus allele counts per locus vary little populations and are therefore less informative. The number of segregating sites per population is a useful alternative in SNP datasets given the large number of possible segregating sites that these datasets often query [@?CITE].

However, segregating sites counts per population are problematic whenever sample sizes are not equal across the populations under comparison at all loci. This can occur either directly due to unequal numbers of individuals or  indirectly due to unqual proportions of missing data across populations. This problem also effects estimates of allele counts per locus and private allele numbers, but corrections for both which use rarefaction to estimate those parameters under a common sample size are well developed [@kalinowskiCountingAllelesRarefaction2004].

Here we present rarefaction-corrected estimators for the probability that any given loci would be observed to be segregating under a reduced sample size and the expected total number of segregating sites across all loci within a population. We show that these estimators are highly accurate via comparison to simulated re-sampling using simulated data. These estimators are currently implemented and available in the `snpR` R package via the function `calc_seg_sites` [@hemstromSnpRUserFriendly2023], where they can be automatically run for all populations either with automatic, per-locus optimized control of target sample size or globally via manual control.

# Methods
<!-- https://link.springer.com/article/10.1023/B:COGE.0000041021.91777.1a#citeas -->

## Probability of observing segregating loci via rarefaction
Rarefaction can be used to estimate the probability of observing a segregating site at a specific locus using much the same framework used to calculate allelic richness. In brief, allelic richness (or the expected number of distinct alleles expected at a given locus under a common sample size $g$ across populations) can be estimated for a given population $j$ by summing the probability of observing each $i$ of $m$ unique alleles using the counts of those alleles in the population $N_{ij}$ and the total sample size in that population $N_{j}$. This is done by comparing the number of possible ways to draw $g$ gene copies without sampling allele $i$ ($\binom{N_{j} - N_{ij}}{g}$) to the total number of possible combinations of gene copies that can be drawn ($\binom{N_{j}}{g}$); the inverse of this ($1 - \frac{\binom{N_{j} - N_{ij}}{g}}{\binom{N_{j}}{g}}$) is therefore the probability of observing allele $i$ in population $N_{j}$, and the sum of this value across all $m$ alleles gives the expected number of alleles observed at a locus in population $j$, $\alpha_{g}^{j}$ [@hurlbertNonconceptSpeciesDiversity1971;@kalinowskiCountingAllelesRarefaction2004]:
$$\alpha_{g}^{j} = \sum_{i = 1}^{m} 1 - \frac{\binom{N_{j} - N_{ij}}{g}}{\binom{N_{j}}{g}}$$

The expected number of segregating loci in a population for a draw of $g$ gene copies can be derived similarly. For a locus $i$ to be segregating in population $j$, all alleles drawn across all gene copies must be identical. If alleles are independent at each locus (the locus is at Hardy-Weinburg Equlibrium, HWE) and $N$ is infinite, the probability ($P({S_{j}})$) of observing a segregating site at a locus is the inverse of the probability of drawing only one allele in $g$ draws with replacement: $$P({S_{j}}) = 1 - \sum_{i = 1}^{m} f_{ij}^{g}$$

where $f_{ij}$ is the allele frequency of allele $i$ in population $j$. However, in finite samples draws are conducted with replacement, and so binomial coefficients must instead be used to determine the probability of drawing only a specific allele:
$$P({S_{j}}) = 1 - \sum_{i = 1}^{m} \frac{\binom{N_{ij}}{g}}{\binom{N_{j}}{g}}$$

However, HWE is often not a desirable assumption to make. Even if filtering is employed to remove loci which do not conform to HWE, the degree of conformity, and thus the degree of statistical bias in estimating $P({S_{j}})$, typically varies somewhat between populations. For example, in a sample of 100 genotypes with a minor allele frequency of 0.05, only five minor alleles are expected and two out of three possible combinations of minor homozygotes and heterozygotes that produce that frequency will not deviate from HWE at $\alpha = 0.05$ according to an exact test [@Wigginton2005]. However, re-sampling these to, say, ten genoytpes should will produce quite different $P(S_{j})$ (roughly 0.7, 0.6, and 0.5, for purely heterozygotes, one homozygote and three heterozygotes, and two homozygotes and one heterozygote, respectively)  as we will see below.

To remedy these problems, we propose the following estimator of $P({S_{j}})$:
$$P({S_{j}}) = 1 - \sum_{k = 1}^{h}\frac{\binom{n_{kj}}{\gamma}}{\binom{n_{j}}{\gamma}}$$
where the $P({S_{j}})$ is given by the probability of exclusively drawing any $k$ of $h$ possible homozygote genotypes in population $j$ given $\gamma$ independent sampled *genotypes* (not *gene copies*) from the pool of observed genotypes. Here, $n_{kj}$ is the number of observed homozygote genotypes of type $k$ in population $j$ and $n_{j}$ is the total number of observed genotypes of all types. Note that $n_{j}$ and $\gamma$ will be half the value of their equivalents $N_{j}$ and $g$ for diploid species, one third for triploids, and so on.

Interestingly, this method, like the richness method and related private allele rarefaction approaches can smoothly account for varying amounts of missing data at specific loci in different populations by varying $g$ or $\gamma$ across loci. Both can be set to one less than the smallest observed $N_{j}$ or $n_{j}$, the highest values at which rarefaction can be applied within a population, across all populations after accounting for missing data, and can thus vary across loci without bias. Setting either value to $N_{j}$ or $n_{j}$ will instead return the observed allele diversity or segregating site status, respectively.

This is particularly useful given that $E(N_{S})$ or the expected total number of segregating sites, is often of specific interest as a measure of genetic diversity when comparing populations. Given that the expected number of segregating sites at a specific locus $q$ in population $j$, $E(N_{S_{jq}})$, is equal to $P({S_{jq}})$, $E(N_{S})$ can be calculated by summing $P({S_{jq}})$ across all $Q$ loci: $$E(N_{S}) = \sum_{q = 1}^{Q} P({S_{jq}})$$
with $\gamma$ set accordingly for each locus. In this case, $0 \le E(N_{S_{jq}}) \le 1$ for all loci (and thus $0 \le E(N_{S}) \le Q$). 

Usefully, under this framework each locus represents a single Bernoulli trial in which it can be observed to be segregating or not with probability $P({S_{jq}})$. As such, the variance of $P({S_{jq}})$ for each locus is given by $$\sigma_{P(S_{jq})}^{2} = P(S_{jq}) \times (1 - P(S_{jq}))$$
and, if each locus is independent, the variance of $E(N_{S})$ is equal to the sum of $\sigma_{P(S_{jq})}^{2}$ across all loci:
$$\sigma_{E(n_{S})}^{2} = \sum_{q = 1}^{Q}\sigma_{P(S_{jq})}^{2}$$

Confidence and prediction intervals can then be derived using standard approaches for the sum of random, independent Bernoulli trials. When $Q$ is large, for example, the distribution of $E(N_{S})$ should be approach normal and confidence and prediction intervals can be derived using standard normal approximation using the equations $$CI_{N_{S}} = E(N_{S}) \pm Z\sqrt{\frac{\sigma_{E(n_{S})}^{2}}{Q}}$$ $$PI_{N_{S}} = E(N_{S}) \pm Z\sqrt{\sigma_{E(n_{S})}^{2}(1+(1/Q))}$$
where $Z$ is given by the normal quantile function $Z = Q_{X}(1-\alpha)$ with $\mu = 0$ and $\sigma = 1$ for a desired confidence level $\alpha$.

## Emperical Validation
To validate equations 4 and 5, we simulated genotypic data for two populations with sizes 250 and 2,500, each with 100 bi-allelic loci with minor allele frequencies spaced equally between 0.01 and 0.1. We added missing data to each population assigning each locus a missing data rate $R_{mq}$ from a uniform distribution such that $R_{mq} \sim U(0,.3)$, ensuring that overall allele frequencies in each population were maintained. We then used the methods described above to estimate $P({S_{jq}})$, $E(N_{S})$, their variances, and 95% confidence and prediction intervals for $\gamma = $ 10, 20, ..., 100. For comparison, we also conducted 1000, 2000, ..., 10,000 random draws for each $\gamma$ from every loci in each population, then calculated $P({S_{jq}})$ and its variance for each locus empirically and $E(N_{S})$ by summing across all 100 loci for each set of draws.

To compare our calculated estimates to the empirical simulations, we used the implementation of the Agresti-Coull [@agrestiApproximateBetterExact1998] method from the R package `binom` [@dorai-rajBinomBinomialConfidence2022] to calculate 95% confidence intervals for each empirical $P({S_{jq}})$ derived from the simulations described above.

# Results
The methods described here for calculating $P({S_{j}})$ and $E(N_{S})$ performed well. Individual $P({S_{jq}})$ values for each locus and $\gamma$ from each population were within the 95% confidence intervals calculated from their respective simulations ~95% of the time, with no substantial bias across minor allele frequency or the number of simulations (Figures 1-2), although confidence intervals calculated from simulations with higher $\gamma$ values tended to contain $P({S_{j}})$ slightly more often (Figure S1). Note that variation along the generally correlated minor allele frequency/$P({S_{jq}})$ axes visible in Figure 1 due to variations in genotype frequencies in the simulated data for a given minor allele frequency. Given that the $P({S_{jq}})$ values estimated using equation 4 track the confidence intervals calculated from simulations, this is accounted for adequately. 

Calculations of the total number of segregating sites after rarefaction were likewise very similar to those observed via simulating random draws for any $\gamma$ or iteration counts. Specifically, $E(N_{S})$ values calculated via Equation 5 were very close to the mean number of segregating sites in each population after rarefaction (Figure 3, Figure S2). 95% prediction intervals were generally very close to observed 95% quantiles from simulations across $\gamma$ and simulation counts but were consistatntly slightly overestimated on both ends. This is expected given that Equation 9 assumes normality; the actual distribution of $N_{S}$ values are slightly non-normal across samples. Specifically, the medians, but not the means, of the observed samples are therefore slightly above $E(N_{S})$ (Figure S2). Both point estimates of the number of segregating sites and 95% prediction intervals are similarly accurate across a wide range of both $\gamma$ values and simulation counts (Figure 4).

# Discussion
We present here an method to correct for the probability that a given loci would be observed as segregating following rarefaction to a given sample size. This provides for the straightforward estimation of the total number of segregating sites which would be observed in a population sample at any reduced size, and therefore provides a way to standardize that metric across samples from different populations. We also show that this approach is unbiased by allele frequency or missing data variation across individual samples or loci.

Our estimator should useful given that the number of segregating sites can provide an excellent counterpoint to measures of genetic diversity that are based on allele frequencies (such as observed and expected heterozygosity). For example, populations that have experienced a recent population expansion will often carry an excess of low frequency variants [@gattepailleInferringPopulationSize2013] caused by the recent increase in the overall rate at which mutations are produced in the population and the relative lack of time for any such new variants to drift to higher frequencies. The average expected heterozygosity across segregating sites may actually be lower (or at least lower than expected) in such cases than populations which have been demographically static, but they will carry far more segregating sites. A recent study in yellow perch provides an excellent example of such: several recently expanded populations show a relatively slight difference in heterozygosity in comparison to other, more demographically static populations but are segregating at far more loci [@schraidtDispersiveCurrentsExplain2023].

In cases like this, it is particularly important that the number of segregating sites is being properly calculated across populations, since failing to correct for differences in sample size or data missingness could mask biologically interesting signals of demographic history and obscure a critical facet of overall genomic diversity. The $E(N_{S})$ method we present here should therefore be a useful for future studies of genetic diversity, and is currently available for use via the function "calc_seg_sites()" in the R package "snpR" [@hemstromSnpRUserFriendly2023].


# References

# Figure Captions:

Figure 1: The expected probabilities  of observing a segregating site at each locus ($P({S_{jq}})$) for loci with different minor allele frequencies. Probabilities are derived from Equation 4 for each locus for population sizes of 250 and 2500 rarefacted to either $N = 10$ or $N = 100$, corresponding to $\gamma = 10$ or $\gamma = 100$. Points vary from the trend lines due to the addition of missin data: each locus has a random, independent percentage of missing data which varies between 0 and 30. Points are colored depending on if they fell within a 95% confidence interval based on 10,000 simulated rarefaction trials of each population to each $N$, which is marked for each point with an error bar.

Figure 2: Trends in difference between the mathematically expected probability that a locus segregates after rarefaction $P({S_{jq}})$ and the observed probability of segregation following simulated rarefaction across different rarefaction sizes ($\gamma$), minor allele frequencies, and number of simulated rarefaction events. Loci are colored depending on if the expected probabilities fell within the observed 95% prediction intervals.  

Figure 3: The distribution of the total number of segregating sites ($N_{S}$) observed for 10,000 replicate simulated rarefaction trials to either $N = 10$ or $N = 100$ (corresponding to either $\gamma = 10$ or $\gamma = 100$, respectively) for starting population sizes of either $N = 250$ or $N = 2500$. The mathematically expected number of segregating sites ($E(N_{S})$) and 95% prediction intervals on that number are shown with solid yellow or dashed blue lines for each for each distribution.

Figure 4: Trends in the distribution of the total number of segregating sites ($N_{S}$) observed following rarefaction trails across a range of rarefaction sizes ($\gamma$) and number of trials for starting population sizes of either $N = 250$ or $N = 2500$. Horizontal lines on each distribution note 95% quantile limits. The mathematically expected number of segregating sites ($E(N_{S})$) and 95% prediction intervals on that number are shown in yellow error bars to the right of each distribution.
