---
title: "manuscript"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  word_document:
    reference_docx: "paper-files/style_ref.docx"
  pdf_document:
    keep_tex: yes
    fig_caption: yes
    latex_engine: pdflatex
    fig_width: 7
    fig_height: 6
  html_document:
    df_print: paged
bibliography: "paper-files/citations.bib"
csl: "paper-files/apa.csl"
header-includes:
- \usepackage[left]{lineno}
- \linenumbers
- \usepackage{setspace}\doublespacing
- \usepackage{sectsty}\sectionfont{\fontsize{12}{12}\selectfont}
- \usepackage{sectsty}\subsectionfont{\normalfont\itshape\fontsize{12}{12}\selectfont}
- \usepackage[round]{natbib}
indent: yes
sansfont: Times New Roman
fontsize: 12pt
always_allow_html: yes
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(snpR); library(data.table); library(ggplot2); library(cowplot)
```

# Abstract

# Introduction

Estimating the degree of genetic diversity has long been a critically important and widespread practice in conservation genetics [@dewoodyLongstandingSignificanceGenetic2021]. Genetic diversity, or the diversity alleles, genotypes, and segregating loci within a population [@frankhamIntroductionConservationGenetics2002], is the both the basic foundation upon which selection acts and thus in part controls population adaptability [@Kardos2021], and is strongly correlated with fitness both at the individual and population level, even when no obvious functional ties are known [reviewed in @dewoodyLongstandingSignificanceGenetic2021]. 

Estimators of population genetic diversity are themselves extremely diverse, but can essentially be split into two categories depending on if they are allele frequency dependent or independent. Allele frequency dependent statistics, such as expected heterozygosity ($H_{e}$), the F-statistics ($F_{IS}$, $F_{ST}$, and $F_{IT}$)[@Weir1984], nucleotide diversity ($\pi$), and Tajima's $\theta$ [@Tajima1989] all rely principally on the frequency of the alleles across any surveyed loci; whereas allele frequency independent statistics, such as Watterson's $\theta$ [@wattersonNumberSegregatingSites1975], counts of alleles per loci, the number of observed segregating sites, and (to some degree) private allele counts do not. This is not a trivial difference:  evolutionary forces do not act equally on the number of loci/alleles in a population and the frequency of those alleles. Tajima's D, for example, is a powerful and broadly used statistic fundamentally based on the difference between Tajima's $\theta$ and Watterson's $\theta$ can be used to both detect selection and population demographic changes [@Tajima1989] due to the way that those forces act on the balance of allele frequencies within populations.

The average allele count per locus (which is usually corrected to allelic richness [@kalinowskiCountingAllelesRarefaction2004]) has historically been one of the most prevalent allele frequency independent measures of genetic diversity. While this is a particularly useful measure in microsatellite and other datasets utilizing heavily poly-allelic markers, it is less so in single-nucleotide polymorphism (SNP) datasets. SNP datasets are usually primarily biallelic, and thus the count of the number of alleles per locus in such datasets is less variable across populations and therefore less interesting. Counts of the number of segregating sites per population are a useful alternative in SNP datasets given the large number of possible segregating sites that these datasets often query [@?CITE].

However, raw counts of the number of segregating sites per population are problematic whenever sample sizes are not equal across the populations under comparison at all loci. This can occur either directly due to unequal numbers of individuals or  indirectly due to unqual proportions of missing data across populations. This problem also effects estimates of allele counts per locus and private allele numbers, but corrections for both which use rarefaction to estimate those parameters under a common sample size are well developed [@kalinowskiCountingAllelesRarefaction2004].

Here we present rarefaction-corrected estimators for the probability that any given loci would be observed to be segregating under a reduced sample size and the expected total number of segregating sites across all loci within a population. We show that these estimators are highly accurate via comparison to simulated re-sampling using simulated data. These estimators are currently implemented and available in the `snpR` R package via the function `calc_seg_sites` [@hemstromSnpRUserFriendly2023], where they can be automatically run for all populations either with automatic, per-locus optimized control of target sample size or globally via manual control.

<!-- Outline: -->

<!-- * What is genetic diveristy? -->
<!--   - Utility -->
<!--   - Estimators -->
<!--     - heterozygosity -->
<!--     - pi -->
<!--     - private alleles -->
<!--     - allele counts per locus -->

<!-- * Problems with ac and private alleles per locus -->
<!--   - All subject to bias due to differences in effective sample size -->
<!--     - Missing data, unequal sample sizes -->

<!-- * Rarefaction -->
<!--   - Sampling vs mathematical solutions -->
<!--   - solutions for private alleles and allele counts (richness) -->
<!--   - description -->
<!--   - how this solves problems -->

<!-- * Nseg -->
<!--   - Newer metric for modern sequencing tech -->
<!--   - Useful alternative, lets you capture different parts of variance -->
<!--   - biases -->
<!--   - rarefaction as solution -->

# Methods
<!-- https://link.springer.com/article/10.1023/B:COGE.0000041021.91777.1a#citeas -->

## Probability of observing segregating loci via rarefaction
Rarefaction can be used to estimate the probability of observing a segregating site at a specific locus using much the same framework used to calculate allelic richness. In brief, allelic richness (or the expected number of distinct alleles expected at a given locus under a common sample size $g$ across populations) can be estimated for a given population $j$ by summing the probability of observing each $i$ of $m$ unique alleles using the counts of those alleles in the population $N_{ij}$ and the total sample size in that population $N_{j}$. This is done by comparing the number of possible ways to draw $g$ gene copies without sampling allele $i$ ($\binom{N_{j} - N_{ij}}{g}$) to the total number of possible combinations of gene copies that can be drawn ($\binom{N_{j}}{g}$); the inverse of this ($1 - \frac{\binom{N_{j} - N_{ij}}{g}}{\binom{N_{j}}{g}}$) is therefore the probability of observing allele $i$ in population $N_{j}$, and the sum of this value across all $m$ alleles gives the expected number of alleles observed at a locus in population $j$, $\alpha_{g}^{j}$ [@hurlbertNonconceptSpeciesDiversity1971;@kalinowskiCountingAllelesRarefaction2004]:
$$\alpha_{g}^{j} = \sum_{i = 1}^{m} 1 - \frac{\binom{N_{j} - N_{ij}}{g}}{\binom{N_{j}}{g}}$$

The expected number of segregating loci in a population for a draw of $g$ gene copies can be derived similarly. For a locus $i$ to be segregating in population $j$, all alleles drawn across all gene copies must be identical. If alleles are independent at each locus (the locus is at Hardy-Weinburg Equlibrium, HWE) and $N$ is infinite, the probability ($P({S_{j}})$) of observing a segregating site at a locus is the inverse of the probability of drawing only one allele in $g$ draws with replacement: $$P({S_{j}}) = 1 - \sum_{i = 1}^{m} f_{ij}^{g}$$

where $f_{ij}$ is the allele frequency of allele $i$ in population $j$. However, in finite samples draws are conducted with replacement, and so binomial coefficients must instead be used to determine the probability of drawing only a specific allele:
$$P({S_{j}}) = 1 - \sum_{i = 1}^{m} \frac{\binom{N_{ij}}{g}}{\binom{N_{j}}{g}}$$

However, HWE is often not a desirable assumption to make. Even if filtering is employed to remove loci which do not conform to HWE, the degree of conformity, and thus the degree of statistical bias in estimating $P({S_{j}})$, typically varies somewhat between populations. For example, in a sample of 100 genotypes with a minor allele frequency of 0.05, only five minor alleles are expected and two out of three possible combinations of minor homozygotes and heterozygotes that produce that frequency will not deviate from HWE at $\alpha = 0.05$ according to an exact test [@Wigginton2005]. However, re-sampling these to, say, ten genoytpes should will produce quite different $P(S_{j})$ (roughly 0.7, 0.6, and 0.5, for purely heterozygotes, one homozygote and three heterozygotes, and two homozygotes and one heterozygote, respectively)  as we will see below.

To remedy these problems, I propose the following estimator of $P({S_{j}})$:
$$P({S_{j}}) = 1 - \sum_{k = 1}^{h}\frac{\binom{n_{kj}}{\gamma}}{\binom{n_{j}}{\gamma}}$$
where the $P({S_{j}})$ is given by the probability of exclusively drawing any $k$ of $h$ possible homozygote genotypes in population $j$ given $\gamma$ independent sampled *genotypes* (not *gene copies*) from the pool of observed genotypes. Here, $n_{kj}$ is the number of observed homozygote genotypes of type $k$ in population $j$ and $n_{j}$ is the total number of observed genotypes of all types. Note that $n_{j}$ and $\gamma$ will be half the value of their equivalents $N_{j}$ and $g$ for diploid species, one third for triploids, and so on.

Interestingly, this method, like the richness method and related private allele rarefaction approaches can smoothly account for varying amounts of missing data at specific loci in different populations by varying $g$ or $\gamma$ across loci. Both can be set to one less than the smallest observed $N_{j}$ or $n_{j}$, the highest values at which rarefaction can be applied within a population, across all populations after accounting for missing data, and can thus vary across loci without bias. Setting either value to $N_{j}$ or $n_{j}$ will instead return the observed allele diversity or segregating site status, respectively.

This is particularly useful given that $E(N_{S})$ or the expected total number of segregating sites, is often of specific interest as a measure of genetic diversity when comparing populations. Given that the expected number of segregating sites at a specific locus $q$ in population $j$, $E(N_{S_{jq}})$, is equal to $P({S_{jq}})$, $E(N_{S})$ can be calculated by summing $P({S_{jq}})$ across all $Q$ loci: $$E(N_{S}) = \sum_{q = 1}^{Q} P({S_{jq}})$$
with $\gamma$ set accordingly for each locus. In this case, $0 \le E(N_{S_{jq}}) \le 1$ for all loci (and thus $0 \le E(N_{S}) \le Q$). 

Usefully, under this framework each locus represents a single Bernoulli trial in which it can be observed to be segregating or not with probability $P({S_{jq}})$. As such, the variance of $P({S_{jq}})$ for each locus is given by $$\sigma_{P(S_{jq})}^{2} = P(S_{jq}) \times (1 - P(S_{jq}))$$
and, if each locus is independent, the variance of $E(N_{S})$ is equal to the sum of $\sigma_{P(S_{jq})}^{2}$ across all loci:
$$\sigma_{E(n_{S})}^{2} = \sum_{q = 1}^{Q}\sigma_{P(S_{jq})}^{2}$$

Confidence and prediction intervals can then be derived using standard approaches for the sum of random, independent Bernoulli trials. When $Q$ is large, for example, the distribution of $E(N_{S})$ should be approach normal and confidence and prediction intervals can be derived using standard normal approximation using the equations $$CI_{N_{S}} = E(N_{S}) \pm Z\sqrt{\frac{\sigma_{E(n_{S})}^{2}}{Q}}$$ $$PI_{N_{S}} = E(N_{S}) \pm Z\sqrt{\sigma_{E(n_{S})}^{2}(1+(1/Q))}$$
where $Z$ is given by the normal quantile function $Z = Q_{X}(1-\alpha)$ with $\mu = 0$ and $\sigma = 1$ for a desired confidence level $\alpha$.

## Emperical Validation
To validate equations 4 and 5, I simulated genotypic data for two populations with sizes 100 and 1000, each with 100 bi-allelic loci with minor allele frequencies spaced equally between 0.01 and 0.1. I added missing data to each population assigning each locus a missing data rate $R_{mq}$ from a uniform distribution such that $R_{mq} \sim U(0,.3)$, ensuring that overall allele frequencies in each population were maintained. I then used the methods described above to estimate $P({S_{jq}})$ and $E(N_{S})$ and their variances given $\gamma = 30$. For comparison, I also conducted 10,000 random draws of size $\gamma$ from each loci in each population, then calculated $P({S_{jq}})$ and its variance for each locus empirically and $E(N_{S})$ by summing across all 100 loci for each set of draws. I likewise calculated the variance of $N_{S}$ directly across all sets of random draws. 

To compare my calculated estimates to the empirical simulations, I used the implementation of the Agresti-Coull [@agrestiApproximateBetterExact1998] method from the R package `binom` [@dorai-rajBinomBinomialConfidence2022] to calculate 95% confidence intervals for each empirical $P({S_{jq}})$. I also used normal approximation to calculate 95% confidence and prediction intervals for $E(N_{S})$ and $N_{S}$ using $\sigma_{E(n_{S})}^{2}$ and the variance directly observed from the simulations, respectively.

# Results
The methods described here for calculating $P({S_{j}})$ and $E(N_{S})$ performed well. Individual $P({S_{jq}})$ values for each locus were within the confidence intervals derived from the simulated values for 97 and 96% of loci from the $n = 100$ and $n = 1000$ populations, respectively (Figure 1). Note that variation along the generally correlated minor allele frequency/$P({S_{jq}})$ axes is due to variations in genotype frequencies in the simulated data for a given minor allele frequency. $P({S_{jq}})$ values estimated using equation 4 account for this adequately.

$E(N_{S})$ values estimated with equation 5 were similarly accurate and were within the 95% confidence intervals produced using the simulated $N_{S}$ for both population sizes (Figure 2). Likewise, the 95% prediction intervals calculated using $\sigma_{E(n_{S})}^{2}$ via equation 7 contained 96.1 and 95.3% of the simulated $N_{S}$ values for $n = 100$ and $n = 1000$, respectively (Figure 2).

# Discussion



# References