---
title: "manuscript"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  word_document:
    reference_docx: "paper-files/style_ref.docx"
  pdf_document:
    keep_tex: yes
    fig_caption: yes
    latex_engine: pdflatex
    fig_width: 7
    fig_height: 6
  html_document:
    df_print: paged
bibliography: "paper-files/citations.bib"
csl: "paper-files/apa.csl"
header-includes:
- \usepackage[left]{lineno}
- \linenumbers
- \usepackage{setspace}\doublespacing
- \usepackage{sectsty}\sectionfont{\fontsize{12}{12}\selectfont}
- \usepackage{sectsty}\subsectionfont{\normalfont\itshape\fontsize{12}{12}\selectfont}
- \usepackage[round]{natbib}
indent: yes
sansfont: Times New Roman
fontsize: 12pt
always_allow_html: yes
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(snpR); library(data.table); library(ggplot2); library(cowplot)
```

# Abstract

# Introduction
<!-- Outline: -->

<!-- * What is genetic diveristy? -->
<!--   - Utility -->
<!--   - Estimators -->
<!--     - heterozygosity -->
<!--     - pi -->
<!--     - private alleles -->
<!--     - allele counts per locus -->

<!-- * Problems with ac and private alleles per locus -->
<!--   - All subject to bias due to differences in effective sample size -->
<!--     - Missing data, unequal sample sizes -->

<!-- * Rarefaction -->
<!--   - Sampling vs mathematical solutions -->
<!--   - solutions for private alleles and allele counts (richness) -->
<!--   - description -->
<!--   - how this solves problems -->

<!-- * Nseg -->
<!--   - Newer metric for modern sequencing tech -->
<!--   - Useful alternative, lets you capture different parts of variance -->
<!--   - biases -->
<!--   - rarefaction as solution -->

# Methods
Outline:

* Background on methods for private alleles, richness
* Solution for segregating sites
  - Derivation
  - Can account for both different sample sizes and missingness across loci and populations.
* Simulations
  - Sampling approach, missingness, etc

<!-- https://link.springer.com/article/10.1023/B:COGE.0000041021.91777.1a#citeas -->

## Probability of observing segregating loci via rarefaction
Rarefaction can be used to estimate the probability of observing a segregating site at a specific locus using much the same framework used to calculate allelic richness. In brief, allelic richness (or the expected number of distinct alleles expected at a given locus under a common sample size $g$ across populations) can be estimated for a given population $j$ by summing the probability of observing each $i$ of $m$ unique alleles using the counts of those alleles in the population $N_{ij}$ and the total sample size in that population $N_{j}$. This is done by comparing the number of possible ways to draw $g$ gene copies without sampling allele $i$ ($\binom{N_{j} - N_{ij}}{g}$) to the total number of possible combinations of gene copies that can be drawn ($\binom{N_{j}}{g}$); the inverse of this ($1 - \frac{\binom{N_{j} - N_{ij}}{g}}{\binom{N_{j}}{g}}$) is therefore the probability of observing allele $i$ in population $N_{j}$, and the sum of this value across all $m$ alleles gives the expected number of alleles observed at a locus in population $j$, $\alpha_{g}^{j}$ [@hurlbertNonconceptSpeciesDiversity1971;@kalinowskiCountingAllelesRarefaction2004]:
$$\alpha_{g}^{j} = \sum_{i = 1}^{m} 1 - \frac{\binom{N_{j} - N_{ij}}{g}}{\binom{N_{j}}{g}}$$

The expected number of segregating loci in a population for a draw of $g$ gene copies can be derived similarly. For a locus $i$ to be segregating in population $j$, all alleles drawn across all gene copies must be identical. If alleles are independent at each locus (the locus is at Hardy-Weinburg Equlibrium, HWE) and $N$ is infinite, the probability ($P({S_{j}})$) of observing a segregating site at a locus is the inverse of the probability of drawing only one allele in $g$ draws with replacement: $$P({S_{j}}) = 1 - \sum_{i = 1}^{m} f_{ij}^{g}$$

where $f_{ij}$ is the allele frequency of allele $i$ in population $j$. However, in finite samples draws are conducted with replacement, and so binomial coefficients must instead be used to determine the probability of drawing only a specific allele:
$$P({S_{j}}) = 1 - \sum_{i = 1}^{m} \frac{\binom{N_{ij}}{g}}{\binom{N_{j}}{g}}$$

However, HWE is often not a desirable assumption to make. Even if filtering is employed to remove loci which do not conform to HWE, the degree of conformity, and thus the degree of statistical bias in estimating $P({S_{j}})$, typically varies somewhat between populations.

To remedy this, I propose the following estimator of $P({S_{j}})$:
$$P({S_{j}}) = 1 - \sum_{k = 1}^{h}\frac{\binom{n_{kj}}{\gamma}}{\binom{n_{j}}{\gamma}}$$
where the $P({S_{j}})$ is given by the probability of exclusively drawing any $k$ of $h$ possible homozygote genotypes in population $j$ given $\gamma$ independent sampled *genotypes* (not *gene copies*) from the pool of observed genotypes. Here, $n_{kj}$ is the number of observed $k$ genotypes in population $j$ where the total number of observed genotypes of all types is $n_{j}$. Note that $n_{j}$ and $\gamma$ will be half the value of their equivalents $N_{j}$ and $g$ for diploid species, one third for triploids, and so on.

Interestingly, this method, like the richness method and related private allele rarefaction approaches can smoothly account for varying amounts of missing data at specific loci in different populations by varying $g$ or $\gamma$ across loci. Both can be set to one less than the smallest observed $N_{j}$ or $n_{j}$, the highest values at which rarefaction can be applied within a population, across all populations after accounting for missing data, and can thus vary across loci without bias. Setting either value to $N_{j}$ or $n_{j}$ will instead return the observed allele diversity or segregating site status, respectively.

This is particularly useful given that $E(N_{S})$ or the expected total number of segregating sites, is often of specific interest as a measure of genetic diversity when comparing populations. Given that the expected number of segregating sites at a specific locus $q$ in population $j$, $E(N_{S_{jq}})$, is equal to $P({S_{jq}})$, $E(N_{S})$ can be calculated by summing $P({S_{jq}})$ across all $Q$ loci: $$E(N_{S}) = \sum_{q = 1}^{Q} P({S_{jq}})$$
with $\gamma$ set accordingly for each locus. In this case, $0 \le E(N_{S_{jq}}) \le 1$ for all loci (and thus $0 \le E(N_{S}) \le Q$). 

Usefully, under this framework each locus represents a single Bernoulli trial in which it can be observed to be segregating or not with probability $P({S_{jq}})$. As such, the variance of $P({S_{jq}})$ for each locus is given by $$\sigma_{P(S_{jq})}^{2} = P(S_{jq}) \times (1 - P(S_{jq}))$$
and, if each locus is independent, the variance of $E(N_{S})$ is equal to the sum of $\sigma_{P(S_{jq})}^{2}$ across all loci:
$$\sigma_{E(n_{S})}^{2} = \sum_{q = 1}^{Q}\sigma_{P(S_{jq})}^{2}$$

Confidence and prediction intervals can then be derived using standard approaches for the sum of random, independent Bernoulli trials. When $Q$ is large, for example, the distribution of $E(N_{S})$ should be approach normal and confidence and prediction intervals can be derived using standard normal approximation. In contrast, confidence intervals for individual $P({S_{jq}})$ values can be achieved via standard approaches for single Bernoulli samples of size $\gamma$.

## Emperical Validation
To validate equations 4 and 5, I simulated genotypic data for two populations with sizes 100 and 1000, each with 100 bi-allelic loci with minor allele frequencies spaced equally between 0.01 and 0.1. I added missing data to each population assigning each locus a missing data rate $R_{mq}$ from a uniform distribution such that $R_{mq} \sim U(0,.3)$, ensuring that overall allele frequencies in each population were maintained. I then used the methods described above to estimate $P({S_{jq}})$ and $E(N_{S})$ and their variances given $\gamma = 30$. I used the "asymptotic" method from the R package `binom` [@rajBinomBinomialConfidence2022] to calculate 95% confidence intervals for each $P({S_{jq}})$ and normal approximation to calculate 95% confidence and prediction intervals for $E(N_{S})$. 

For comparison, I also conducted 10,000 random draws of size $\gamma$ from each loci in each population, then calculated $P({S_{jq}})$ and its variance for each locus empirically and $E(N_{S})$ by summing across all 100 loci for each set of draws. I likewise calculated the variance of $E(N_{S})$ directly across all sets of random draws. 

# Results
The methods described here for calculating $P({S_{j}})$ and $E(N_{S})$ performed well. 95% confidence intervals for simulated loci overlapped the simulated mean in 