---
title: "Seg_sites_rarefaction"
output: html_document
date: "2023-08-11"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE); library(snpR); library(ggplot2); library(data.table);
library(cowplot)
```


# Define parameters
```{r}
# define parameters
seed <- 231266
threads <- 25
allele_frequencies <- seq(.001, .1, by = .001) # allele frequencies
n1 <- 5000 # p1 gene copy number
n2 <- 500 # p2 gene copy number
k <- seq(10, 100, by = 10) # numbers to rarefact to
mcmcs <- seq(1000, 10000, by = 1000) # number of mcmc draws
max_missing <- 30 # maximum percentage of missing data in a locus, from a uniform dist


## derived
set.seed(seed)
nal <- length(allele_frequencies)
miss_percent <- floor(runif(nal, 0, max_missing))/100

## plotting
colors <- khroma::color("highcontrast")(3)
ci_col <- as.character(colors[3])
colors <- as.character(colors[c(2,1)])
t.sizes <- c(20, 18)
```


# Setup genotypes
```{r}
# generate genotypes
g1 <- matrix(0, nrow = nal, ncol = n1) # init
g2 <- matrix(0, nrow = nal, ncol = n2)

# add missing data
ms1 <- lapply(n1*miss_percent, function(x) sample(n1, x, FALSE))  # which are missing
ms2 <- lapply(n2*miss_percent, function(x) sample(n2, x, FALSE))

# determine minor allele placements
ma1 <- ma2 <- vector("list", length(ms1))
for(i in 1:nal){
  ma1[[i]] <- sample(n1*(1-miss_percent)[i], (n1*(1-miss_percent)*allele_frequencies)[i], FALSE)
  ma2[[i]] <- sample(n2*(1-miss_percent)[i], (n2*(1-miss_percent)*allele_frequencies)[i], FALSE)

}

# populate
## function to replace minors
populate <- function(g, missing, minors){
  if(length(missing) > 0){
    g[missing] <- NA
    g[-missing][minors] <- 1
  }
  else{
    g[minors] <- 1
  }
  return(g)
}
## do for every locus
for(i in 1:nal){ 
  g1[i,] <- populate(g1[i,], ms1[[i]], ma1[[i]])
  g2[i,] <- populate(g2[i,], ms2[[i]], ma2[[i]])
}

## combine into individuals
g1 <- g1[,1:(ncol(g1)/2)] + g1[,((ncol(g1)/2) + 1):ncol(g1)]
g2 <- g2[,1:(ncol(g2)/2)] + g2[,((ncol(g2)/2) + 1):ncol(g2)]

```

# Rarefact mathematically
```{r}
counts <- function(g){
  return(data.frame(majhom = rowSums(g == 0, na.rm = TRUE), 
                    het = rowSums(g == 1, na.rm = TRUE), 
                    minhom = rowSums(g == 2, na.rm = TRUE)))
}
rarefact <- function(counts, k){
  if(length(k) > 1){
    warning("length(k) > 1, using first element only.\n")
    k <- k[1]
  }
  return(1 - ((choose(counts[,1], k) + choose(counts[,3], k))/choose(rowSums(counts), k)))
}
varseg <- function(p) p*(1-p)

c1 <- counts(g1) # geno counts
c2 <- counts(g2)

# rarefact
pseg1 <- lapply(k, function(tk) rarefact(c1, tk))
pseg2 <- lapply(k, function(tk) rarefact(c2, tk))
names(pseg1) <- names(pseg2) <- paste0("K_", k)

vseg1 <- lapply(pseg1, varseg)
vseg2 <- lapply(pseg2, varseg)

p1 <- lapply(pseg1, function(p) sum(p)/nal)
p2 <- lapply(pseg2, function(p) sum(p)/nal)
```


# Rarefact via simulations
Note: this chunck should run on a cluster or somewhere else allowing for large numbers of threads, then be re-imported in the next chunck. `q` should be passed as an argument for the iteration, as should `outfile`.

```{r cluster_MCMCs, eval=FALSE}
set.seed(seed + q) # reset seed for cluster

rarefact_mcmc <- function(g, k){
  if(length(k) > 1){
    warning("length(k) > 1, using first element only.\n")
    k <- k[1]
  }
  
  miss <- is.na(g)
  if(any(miss)){
    g <- g[-which(miss)]
  }
  
  samp <- sample(g, k, FALSE)
  seg <- ifelse(any(samp == 1), 1, # got hets
                ifelse(any(samp == 0) & any(samp == 2), 1, 0)) # got both homs
  return(seg)
}

# draw for each k/mcmc count combination
combs <- expand.grid(k, mcmcs)
colnames(combs) <- c("k", "iters")

tmcmcs <- combs$iters[q]
tk <- combs$k[q]

seg1 <- seg2 <- matrix(0, nal, tmcmcs)
for(i in 1:tmcmcs){
  for(j in 1:nal){
    seg1[j,i] <- rarefact_mcmc(g1[j,], k = tk)
    seg2[j,i] <- rarefact_mcmc(g2[j,], k = tk)
  }
}

mpseg1 <- binom::binom.confint(rowSums(seg1), rep(tmcmcs, nrow(seg1)),
                               method = "agresti-coull")[,c("mean", "lower", "upper")]
mpseg2 <- binom::binom.confint(rowSums(seg2), rep(tmcmcs, nrow(seg1)),
                               method = "agresti-coull")[,c("mean", "lower", "upper")]

colnames(mpseg1) <- colnames(mpseg2) <- c("mean", "lower", "upper")

res <- list(seg1 = seg1, seg2 = seg2, mpseg1 = mpseg1, mpseg2 = mpseg2, k = tk, iters = tmcmcs)

saveRDS(res, paste0(outfile, "_", q, ".RDS"))

```

# import and condense simulation results
```{r}
res_files <- list.files("../data/", "\\.RDS", full.names = TRUE)
res_files <- lapply(res_files, readRDS)
ks <- unlist(purrr::map(res_files, "k"))
n_iters <- unlist(purrr::map(res_files, "iters"))

```


# Figure 1: Pseg for four individual runs (k = 10 and 100, N = 100, N = 1000)

```{r Figure1}
prep_f1 <- function(pseg1, pseg2, mpseg1, mpseg2, n1, n2){
  #========plot the emperical CIs around every loci vs the predicted value==============
  # plotting the emperical CIs vs the estiamted mean
  pseg <- rbind(data.frame(pop = paste0("N = ", n1/2), p = pseg1, snp = 1:nal, af = allele_frequencies),
                data.frame(pop = paste0("N = ", n2/2), p = pseg2, snp = 1:nal, af = allele_frequencies))
  
  oseg <- rbind(data.frame(pop = paste0("N = ", n1/2), p = mpseg1[,1], snp = 1:nal, af = allele_frequencies),
                data.frame(pop = paste0("N = ", n2/2), p = mpseg2[,1], snp = 1:nal, af = allele_frequencies))
  
  pseg$source <- "Mathematical"
  oseg$source <- "Simulation"
  
  seg <- rbind(pseg, oseg)
  cis <- rbind(cbind.data.frame(mpseg1, pop = paste0("N = ", n1/2), af = allele_frequencies),
               cbind.data.frame(mpseg2, pop = paste0("N = ", n2/2), af = allele_frequencies))
  colnames(cis) <- c("p", "lower", "upper", "pop", "af")
  seg <- merge(seg, cis[,c("lower", "upper", "af", "pop")], by = c("af", "pop"))
  seg$in_ci <- ifelse(seg$p >= seg$lower & seg$p <= seg$upper, TRUE, FALSE)
  
  return(seg)
}

K10_data <- which(ks == 10 & n_iters == 10000)
F1data_K10 <- prep_f1(pseg1 = pseg1$K_10, 
                      pseg2 = pseg2$K_10, 
                      mpseg1 = res_files[[K10_data]]$mpseg1,
                      mpseg2 = res_files[[K10_data]]$mpseg2,
                      n1 = n1,
                      n2 = n2)

K100_data <- which(ks == 100 & n_iters == 10000)
F1data_K100 <- prep_f1(pseg1 = pseg1$K_100, 
                       pseg2 = pseg2$K_100, 
                       mpseg1 = res_files[[K100_data]]$mpseg1,
                       mpseg2 = res_files[[K100_data]]$mpseg2,
                       n1 = n1,
                       n2 = n2)


F1data_K10$K <- "gamma * ' = 10'"
F1data_K100$K <- "gamma * ' = 100'"

F1data <- rbind(F1data_K10, F1data_K100)
F1data$pop <- paste0("'", F1data$pop, "'")

f1 <- ggplot(F1data[F1data$source == "Mathematical",],
             aes(x = af, y = p)) +
  geom_point(aes(color = in_ci)) +
  geom_errorbar(aes(ymax = upper, ymin = lower)) +
  facet_grid(K~pop, labeller = label_parsed) +
  theme_bw() +
  theme(strip.background = element_blank()) +
  scale_color_manual(values = colors) +
  guides(color = guide_legend(title = "In CI")) +
  ylab(bquote(italic(P*"("*S[jq]*")"))) +
  xlab("Minor Allele Frequency") +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        legend.title = element_text(size = t.sizes[1]),
        legend.text = element_text(size = t.sizes[2]))

ggsave("Figure1.pdf", f1, width = 15, height = 10)
tapply(F1data[F1data$source == "Mathematical", "in_ci"], F1data[F1data$source == "Mathematical", c("K", "pop")], mean) # percent in interval
```

# Figure 2: Distribution of Nseg vs Eseg for four runs

```{r}
prep_f2 <- function(seg1, seg2, vseg1, vseg2, F1data, n1, n2){
  seg <- F1data
  
  # normal approximation CI and PI equation given var and N
  nap_pi <- function(alpha, vseg, n) qnorm(1 - (alpha/2))*sqrt(vseg)*sqrt(1 + (1/n))
  nap_ci <- function(alpha, vseg, n) qnorm(1 - (alpha/2))*sqrt(vseg)/sqrt(n)
  
  # sums across loci
  msum1 <- colSums(seg1)
  msum2 <- colSums(seg2)
  
  # CI/PI
  msum_ci1 <- nap_ci(.05, var(msum1), length(msum1))
  msum_ci2 <- nap_ci(.05, var(msum2), length(msum2))
  
  sum_pi1 <- nap_pi(0.05, sum(vseg1), nal)
  sum_pi2 <- nap_pi(0.05, sum(vseg2), nal)
  
  # add to table
  msum_tab <- data.frame(nseg = c(mean(msum1), mean(msum2)),
                         lower = c(mean(msum1) - msum_ci1, mean(msum2) - msum_ci2),
                         upper = c(mean(msum1) + msum_ci1, mean(msum2) + msum_ci2),
                         pop = c(paste0("n = ", n1/2), paste0("n = ", n2/2)))
  
  sum_tab <- tapply(seg[seg$source == "Mathematical",]$p, seg[seg$source == "Mathematical",]$pop, sum)
  sum_tab <- as.data.frame(sum_tab)
  sum_tab$pop <- tolower(rownames(sum_tab))
  colnames(sum_tab)[1] <- "nseg"
  sum_tab <- merge(sum_tab, msum_tab[,c("pop", "lower", "upper")], by = "pop")
  sum_tab$in_ci <- sum_tab$nseg <= sum_tab$upper & sum_tab$nseg >= sum_tab$lower
  
  # add pi
  all_mcmc_sums <- rbind(cbind.data.frame(nseg = msum1, pop = paste0("n = ", n1/2)),
                         cbind.data.frame(nseg = msum2, pop = paste0("n = ", n2/2)))
  pi_tab <- data.frame(pop = c(paste0("n = ", n1/2), paste0("n = ", n2/2)),
                       lower = c(sum_tab$nseg[sum_tab$pop == paste0("n = ", n1/2)] - sum_pi1,
                                 sum_tab$nseg[sum_tab$pop == paste0("n = ", n2/2)] - sum_pi2),
                       upper = c(sum_tab$nseg[sum_tab$pop == paste0("n = ", n1/2)] + sum_pi1,
                                 sum_tab$nseg[sum_tab$pop == paste0("n = ", n2/2)] + sum_pi2))
  all_mcmc_sums <- merge(all_mcmc_sums, pi_tab, by = "pop")
  all_mcmc_sums$in_pi <- all_mcmc_sums$nseg <=  all_mcmc_sums$upper & all_mcmc_sums$nseg >=  all_mcmc_sums$lower
  
  return(list(sum_tab = sum_tab, all_mcmc_sums = all_mcmc_sums, pi_tab = pi_tab))
}

K10_data <- which(ks == 10 & n_iters == 10000)
F2data_K10 <- prep_f2(seg1 = res_files[[K10_data]]$seg1, 
                      seg2 = res_files[[K10_data]]$seg2, 
                      vseg1 = vseg1$K_10,
                      vseg2 = vseg2$K_10,
                      F1data = F1data_K10,
                      n1 = n1,
                      n2 = n2)

K100_data <- which(ks == 100 & n_iters == 10000)
F2data_K100 <- prep_f2(seg1 = res_files[[K100_data]]$seg1, 
                       seg2 = res_files[[K100_data]]$seg2, 
                       vseg1 = vseg1$K_100,
                       vseg2 = vseg2$K_100,
                       F1data = F1data_K100,
                       n1 = n1,
                       n2 = n2)


F2data_K10$sum_tab$K <- "gamma * ' = 10'"
F2data_K100$sum_tab$K <- "gamma * ' = 100'"
F2data_K10$all_mcmc_sums$K <- "gamma * ' = 10'"
F2data_K100$all_mcmc_sums$K <- "gamma * ' = 100'"
F2data_K10$pi_tab$K <- "gamma * ' = 10'"
F2data_K100$pi_tab$K <- "gamma * ' = 100'"

F2data_sums <- rbind(F2data_K10$sum_tab, F2data_K100$sum_tab)
F2data_sums$pop <- gsub("n", "N", F2data_sums$pop)
F2data_sums$pop <- paste0("'", F2data_sums$pop, "'")

F2data_mcmcs <- rbind(F2data_K10$all_mcmc_sums, F2data_K100$all_mcmc_sums)
F2data_mcmcs$pop <- gsub("n", "N", F2data_mcmcs$pop)
F2data_mcmcs$pop <- paste0("'", F2data_mcmcs$pop, "'")

F2data_pi <- rbind(F2data_K10$pi_tab, F2data_K100$pi_tab)
F2data_pi$pop <- gsub("n", "N", F2data_pi$pop)
F2data_pi$pop <- paste0("'", F2data_pi$pop, "'")


f2 <- ggplot(F2data_mcmcs, aes(x = nseg)) +
  geom_density() +
  geom_vline(data = F2data_sums, aes(xintercept = nseg), color = colors[1]) +
  geom_vline(data = F2data_pi, aes(xintercept = lower), linetype = "dashed", color = colors[2]) +
  geom_vline(data = F2data_pi, aes(xintercept = upper), linetype = "dashed", color = colors[2]) +
  facet_grid(K~pop, labeller = label_parsed) +
  theme_bw() +
  scale_y_continuous(expand = c(0,0)) +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        strip.background = element_blank()) +
  xlab(bquote(italic(N[S]))) +
  ylab("Density")

ggsave("Figure2.pdf", f2, height = 10, width = 15)

```

# Figure 3: var vs difference between estimate and mean of simulations for mcmc and k

```{r Figure3}
# for all of the mpseg files, figure out if the percentage of points in the CI changes
## do for both each SNP and for Ns
ps_list <- ns_list <- vector("list", length(res_files))
for(i in 1:length(res_files)){
  tk <- res_files[[i]]$k
  titers <- res_files[[i]]$iters
  
  ps_prep <- prep_f1(pseg1 = pseg1[[paste0("K_", tk)]],
                     pseg2 = pseg2[[paste0("K_", tk)]], 
                     mpseg1 = res_files[[i]]$mpseg1,
                     mpseg2 = res_files[[i]]$mpseg2,
                     n1 = n1,
                     n2 = n2)
  
  ns_prep <- prep_f2(seg1 = res_files[[i]]$seg1, 
                     seg2 = res_files[[i]]$seg2, 
                     vseg1 = vseg1[[paste0("K_", tk)]],
                     vseg2 = vseg2[[paste0("K_", tk)]],
                     F1data = ps_prep,
                     n1 = n1,
                     n2 = n2)
  
  ps_prep$K <- tk
  ps_prep$iters <- titers
  ps_prep$sim_num <- i
  
  ns_prep$all_mcmc_sums$K <- tk
  ns_prep$all_mcmc_sums$iters <- titers
  ns_prep$all_mcmc_sums$i <- i
  ns_prep$all_mcmc_sums$ENs <- ns_prep$sum_tab$nseg[match(ns_prep$all_mcmc_sums$pop,
                                                          ns_prep$sum_tab$pop)]
  
  ps_list[[i]] <- ps_prep
  ns_list[[i]] <- ns_prep$all_mcmc_sums
}

ps_all <- rbindlist(ps_list)

ps_all_m <- ps_all[source == "Mathematical",]
ps_all_s <- ps_all[source == "Simulation",]
ps_all_merge <- merge(ps_all_m[,-c("source")], 
                      ps_all_s[,-c("source")], 
                      by = c("af", "pop", "snp", "sim_num", "lower", "upper", "K", "iters"), 
                      suffixes = c(".mathematical", ".simulated"))
ps_all_merge[,diff := p.mathematical - p.simulated]

# Figure S1: gamma vs error rate
ps_all_aggregate_for_K <- ps_all_merge[, lapply(.SD, mean), .SDcols = "in_ci.mathematical", by = .(K, iters, pop)]
FS1 <- ggplot(ps_all_aggregate_for_K, aes(x = K, y = in_ci.mathematical*100)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE) +
  facet_grid(iters~pop) +
  theme_bw() +
  # khroma::scale_color_batlow(discrete = TRUE, reverse = TRUE) +
  # khroma::scale_fill_batlow(discrete = TRUE, reverse = TRUE) +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        strip.background = element_blank(),
        legend.title = element_text(size = t.sizes[1]),
        legend.text = element_text(size = t.sizes[2])) +
  xlab(parse(text = "gamma")) +
  ylab("% estimates in CI")

# Figure S2: maf vs error rate
ps_all_aggregate_for_af <- ps_all_merge[, lapply(.SD, mean), .SDcols = "in_ci.mathematical", by = .(af, iters, pop)]
FS2 <- ggplot(ps_all_aggregate_for_af, aes(x = af, y = in_ci.mathematical*100)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE) +
  facet_grid(iters~pop) +
  theme_bw() +
  # khroma::scale_color_batlow(discrete = TRUE, reverse = TRUE) +
  # khroma::scale_fill_batlow(discrete = TRUE, reverse = TRUE) +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        strip.background = element_blank(),
        legend.title = element_text(size = t.sizes[1]),
        legend.text = element_text(size = t.sizes[2])) +
  xlab("Minor Allele Frequency") +
  ylab("% estimates in CI")


f3a <- ggplot(ps_all_merge, aes(x = K, y = diff, color = in_ci.mathematical)) +
  geom_point() +
  theme_bw() +
  facet_grid(~pop) +
  khroma::scale_color_highcontrast() +
  scale_x_continuous(breaks = k) +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        strip.background = element_blank(),
        legend.title = element_text(size = t.sizes[1]),
        legend.text = element_text(size = t.sizes[2])) +
  xlab(parse(text = "gamma")) +
  ylab(bquote(italic(P*"("*S[jq]*")"[E]) - italic(P*"("*S[jq]*")"[O]))) +
  guides(color = guide_legend(title = "In 95% CI"))

f3b <- ggplot(ps_all_merge, aes(x = af, y = diff, color = in_ci.mathematical)) +
  geom_point() +
  theme_bw() +
  facet_grid(~pop) +
  khroma::scale_color_highcontrast() +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        axis.text.x = element_text(size = t.sizes[2], angle = 90, vjust = .5),
        strip.background = element_blank(),
        legend.title = element_text(size = t.sizes[1]),
        legend.text = element_text(size = t.sizes[2])) +
  xlab("Minor Allele Frequency") +
  ylab(bquote(italic(P*"("*S[jq]*")"[E]) - italic(P*"("*S[jq]*")"[O])))

f3c <- ggplot(ps_all_merge, aes(x = iters, y = diff, color = in_ci.mathematical)) +
  geom_point() +
  theme_bw() +
  facet_grid(~pop) +
  khroma::scale_color_highcontrast() +
  scale_x_continuous(breaks = mcmcs) +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        axis.text.x = element_text(size = t.sizes[2], 
                                   angle = 90, vjust = .5),
        strip.background = element_blank(),
        legend.title = element_text(size = t.sizes[1]),
        legend.text = element_text(size = t.sizes[2])) +
  xlab("Number of Simulated Draws") +
  ylab(bquote(italic(P*"("*S[jq]*")"[E]) - italic(P*"("*S[jq]*")"[O])))

f3l <- ggpubr::get_legend(f3a)
f3 <- plot_grid(f3a + guides(color = "none"), 
          f3b + guides(color = "none"), 
          f3c + guides(color = "none"),
          ncol = 1, 
          align = "bl", 
          axis = "bl")

f3 <- plot_grid(f3, f3l, nrow = 1, rel_widths = c(1, .2))

save_plot("Figure3.pdf", f3, base_height = 11, base_width = 15)
```
# Figure 4 Nseg vs parameters

```{r Figure4}
ns_all <- rbindlist(ns_list)

ns_all$diff <- ns_all$ENs - ns_all$nseg

f4a <- ggplot(ns_all[iters == 10000,], aes(x = K, y = nseg)) +
  geom_errorbar(data = unique(ns_all[iters == 10000, c("pop", "K", "iters", "i", "lower", "upper")]),
            aes(x = K, ymin = lower, ymax = upper), color = "#DDAA33", size = 1, 
            position = position_nudge(x = 3.5),
            inherit.aes = FALSE, width = 2) +
  stat_ydensity(aes(group = K), draw_quantiles = c(.025, .975), scale = "width", bw = .4,
                fill = NA, width = 6) +
  facet_wrap(~pop) +
  theme_bw() + 
  scale_x_continuous(breaks = k) +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        legend.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        strip.background = element_blank(),
        legend.text = element_text(size = t.sizes[2])) +
  ylab(bquote(italic(E*"("*N[S]*")") - italic(N[S,observed]))) +
  xlab(bquote(gamma))

f4b <- ggplot(ns_all[K == 50,], aes(x = iters, y = nseg)) +
  geom_errorbar(data = unique(ns_all[K == 50,c("pop", "K", "iters", "i", "lower", "upper")]),
            aes(x = iters, ymin = lower, ymax = upper), color = "#DDAA33", size = 1, 
            position = position_nudge(x = 350),
            inherit.aes = FALSE, width = 200) +
  stat_ydensity(aes(group = iters), draw_quantiles = c(.025, .975), scale = "width", bw = .4,
                fill = NA, width = 600) +
  facet_wrap(~pop) +
  theme_bw() + 
  scale_x_continuous(breaks = mcmcs) +
  theme(axis.text = element_text(size = t.sizes[2]),
        axis.title = element_text(size = t.sizes[1]),
        legend.title = element_text(size = t.sizes[1]),
        strip.text = element_text(size = t.sizes[1]),
        strip.background = element_blank(),
        legend.text = element_text(size = t.sizes[2])) +
  ylab(bquote(italic(E*"("*N[S]*")") - italic(N[S,observed]))) +
  xlab("Number of Simulated Draws")


```


```{r}




#===============plot all of the observed values vs the estimated value for the sums=============
# plot the estimated prediction interval vs the observations and the confint from the simultations




#==========others==================

# variances, per locus
vars <- rbind(cbind.data.frame(estimated = vseg1, 
                               observed = matrixStats::rowVars(seg1),
                               pop = paste0("n = ", n1/2),
                               af = allele_frequencies),
              cbind.data.frame(estimated = vseg2, 
                               observed = matrixStats::rowVars(seg2),
                               pop = paste0("n = ", n2/2),
                               af = allele_frequencies))
ggplot(vars, aes(x = observed, y = estimated, color = af)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1) +
  facet_wrap(~pop) +
  theme_bw() +
  khroma::scale_color_batlow()

# variances, overall
overall_vars <- rbind(cbind.data.frame(estimated = sum(vseg1),
                                       observed = var(msum1),
                                       pop = paste0("n = ", n1/2)),
                      cbind.data.frame(estimated = sum(vseg2),
                                       observed = var(msum2),
                                       pop = paste0("n = ", n2/2)))
```


```{r}
# show difference in HWE divergence with different combinations of 5 minor alleles in 100 genotypes
exd <- rbind(c(rep(0, 45), rep(1, 5)),
             c(rep(0, 46), 2, rep(1, 3)),
             c(rep(0, 47), rep(2, 2), 1))
exds <- import.snpR.data(as.data.frame(exd), snp.meta = data.frame(snp = 1:3), mDat = -9)
exds <- calc_hwe(exds)
get.snpR.stats(exds, stats = "hwe")

exc <- counts(exd)
rarefact(exc, 10)
```



